egui/src/widgets/text_edit.rs

```diff
@@ -741,7 +761,13 @@ fn byte_index_from_char_index(s: &str, char_index: usize) -> usize {
     s.len()
 }
 
-fn insert_text(ccursor: &mut CCursor, text: &mut String, text_to_insert: &str) {
+fn insert_text(
+    ccursor: &mut CCursor,
+    text: &mut String,
+    prev_text: &mut String,
+    text_to_insert: &str,
+) {
+    *prev_text = text.clone();
```

Comment:
```suggestion
    prev_text: &mut Option<String>,
    text_to_insert: &str,
) {
    *prev_text = prev_text.or_else(|| text.clone());
```

This will make it keep the oldest value if there are multiple changes during the frame. The oldest + latest value is what we we want to emit each frame there is a change.

Reply:
This means changing `prev_text: &mut String` to `prev_text: &mut Option<String>` in all the arguments.

Reply:
Again, not sure how I feel about this one. If text is never optional in a `TextEdit` then wrapping it in an `Option` adds extra layers of indirection. If an `EditText` contains "foo" and isn't changed in a frame, then `prev_text` isn't `None`, it's "foo".

Reply:
The goal is that `prev_text` can be `None` when there is no change. This is both a useful signal, and also there is no need to clone the text if there is no change.

In the current code `prev_text == ""` can mean "no change" or it can mean "there was a change, and the text field started empty".

With `Option<String>` `None` means "no change", `Some("")` means "there was a change, and the text field started empty"

Reply:
I certainly won't pick this hill to die on, but I do strongly feel that we should keep these as plain strings for now. I take your point about the extra `clone` and am certainly willing to revisit if there are performance issues, but the alternative is a bunch of checks/unwraps anywhere that needs access to both strings, and I'm not sure that would be any better than a quick clone of what would likely only be a few bytes. I'll make the change if you insist, though.

Your point about `None` being a useful signal is well-taken, though. I did an equality check for `ValueChanged`, and if they're equal then I set the previous value to `None`. Again, if we actually find string equality checks to be a bottleneck then I'll change that to `Option`.

Reply:
I'm a bit confused though.

I agree the simpler solution would be to have `let prev_text = text.clone()` at the start of the function, and then compare with that later. Super-simple, though it costs an extra clone. And no need for any book-keeping of when the text changed. If you want to do this, go ahead (we can optimize it later).

But that is now what the PR currently does - instead it passes in a `mut` ref to `prev_text` into all functions that modify text, and manually set it when the text changes. This avoids the clone when there is no change, but it also has bugs: if I press two keys the same frame, the `prev_text` is set to the second value, not the first initial value. So if the start is `"A"` and I press `B` and then `C`, the current code will change `prev_text` first to `"A"` and then to `"AB"`, and the event will emit `Text changed from "AB" to "ABC"` when it in fact changed from "A" to "ABC".

So the code in the PR has all the the book-keeping to track of when the text changes (which is great for avoid clones) but because it isn't using `None` it has this bug.

Reply:
I guess I'm a little confused too. Is this something you're able to duplicate, or hypothetical?

I'm not the fastest typist, but I placed my fingers over the f and d keys and tried to bring them down simultaneously. I think I succeeded about as well as I could, but either I didn't manage multiple keypresses in a single frame or I received two events. Tried multiple times. I am on a wireless keyboard so perhaps that introduced some delay, but it types quickly enough.

Also, in practical terms, I wouldn't be able to hear two spoken characters at a typical 1/60 framerate.

I'll accept that this isn't technically correct, but in practice it works, and I still maintain that the intent is clearer than it'd be with bunches of unwraps. Your call, though.

bevy_egui_a11y updated to master. Sorry for the delay, stepped away to work on other projects for a bit.

Reply:
Since you are cloning the string each frame, there is no need to pass in `&mut prev_string` to all the functions. cherry-pick this and see: https://github.com/emilk/egui/commit/77b3a154b4a10e8cc9be674a7199fa29380e13bb

Reply:
OK, that does indeed work. Thanks for that--things are still a bit unintuitive but they're starting to click.